### ISSUE TITLE
Get Generate code compilation error before calling AddSource

### ISSUE THREAD

**Issue by edwin-alvarez** [dotnet/roslyn#56294]
**Brief description:**

I want to be able to use .GetCompilationUnitRoot().Errors().Length from my generator before I call AddSource, so I can validate if there are compilation errors. the problem I see is that Errors() is defined under namespace Microsoft.CodeAnalysis.CSharp.UnitTests

**Languages applicable:**

C# only

**Code example that the analyzer should report:**
```

 var scr = GetSyntaxTree(code);
                if((scr.GetCompilationUnitRoot().Errors().Length>0))
                {
                    //trigger error
                }

        private SyntaxTree GetSyntaxTree(string code, params string[] defines)
        {
            var scr = SourceText.From(code, Encoding.UTF8);
            return  SyntaxFactory.ParseSyntaxTree(scr, GetOptions(defines));
        }

        private CSharpParseOptions GetOptions(string[] defines)
        {
            return new CSharpParseOptions(languageVersion: LanguageVersion.CSharp3, preprocessorSymbols: defines);
        }
```

**Comment by jmarolf** [comment]
@edwin-alvarez is there a reason `GetDiagnostics` does not work for your purpose?

```csharp
var scr = GetSyntaxTree(code);
if ((scr.GetCompilationUnitRoot().GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error))
{
    //trigger error
}

private SyntaxTree GetSyntaxTree(string code, params string[] defines)
{
    var scr = SourceText.From(code, Encoding.UTF8);
    return SyntaxFactory.ParseSyntaxTree(scr, GetOptions(defines));
}

private CSharpParseOptions GetOptions(string[] defines)
{
    return new CSharpParseOptions(languageVersion: LanguageVersion.CSharp3, preprocessorSymbols: defines);
}
```

**Comment by edwin-alvarez** [comment]
I couldnt find any example, the only one I found was in the testsuit, but this works great. thanks
